rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,355 @@
+Teleport cluster. At certain points in time, the database schema is read and
+passed through the import rules. The individual import rules may apply custom
+labels. An database object resource will only be created if there is at least
+one import rule that matched it.
+
+Only object attributes (standard ones like `protocol` or custom from the
+`attributes` field), which are sourced from the object spec, are subject to
+matching against the import rules. Any labels present on an object, either from
+an import rule or another source, are not subject to be matched by the import
+rule. The permissions matching is different in this regard.
+
+Example: the `sales-prod` import rule, which imports all tables from `sales`
+Postgres database in prod:
+
+```yaml
+kind: db_object_import_rule
+metadata:
+  name: sales-prod
+spec:
+  db_labels:
+    env: prod
+  mappings:
+    - object_match:
+        - database: sales
+          object_kind: table
+          protocol: postgres
+      add_labels:
+        env: prod
+        product: sales
+  priority: 10
+version: v1
+```
+
+Example database object, with applied labels:
+
+```yaml
+kind: db_object
+metadata:
+  labels:
+    env: prod
+    product: sales
+  name: sales_main
+spec:
+  attributes:
+    attr1: custom attr1 value
+    attr2: custom attr2 value
+  database: sales
+  name: sales_main
+  object_kind: table
+  protocol: postgres
+  schema: public
+  service_name: sales-prod-123
+version: v1
+```
+
+The permissions to particular objects are defined in a role using the new field
+`db_permissions`. The permission specifies the object kind (e.g. `table`),
+permission to grant/revoke (`SELECT`) as well as labels to be matched.
+
+The matching is performed against database object:
+
+- attributes: standard and custom, sourced from the object spec, provided by
+  database-specific schema import code
+- resource labels: aside from standard labels, these can be manipulated by the
+  import rules.
+
+If both attributes and labels provide the same non-empty key, the attributes
+take preference.
+
+Example role `db-dev`, which grants `SELECT` permission to some objects, and
+revokes `UPDATE` from all others.
+
+```yaml
+kind: role
+metadata:
+  name: db-dev
+spec:
+  allow:
+    db_permissions:
+      - match:
+          product: sales
+          protocol: postgres
+        object_kind: table
+        permission: SELECT
+  deny:
+    db_permissions:
+      - object_kind: table
+        permission: UPDATE
+```
+
+The database objects are imported during the connection, to be used for
+permission calculation.
+
+Additionally, the imports will be done on a predetermined schedule (e.g. every
+10 minutes), and stored in the backend.
+
+The database objects stored in the backend will be used by TAG.
+
+The permissions will be applied to the user after the user is provisioned in the
+database. After the session is finished, the user is removed/deactivated, and
+all permissions are revoked.
```

Comment:
Are these permissions applied to the db user on top of existing `db_roles` from auto-user provisioning, or are `db_permissions` and `db_roles` mutuality exclusive?

Reply:
On top, or rather side-by-side. There will be use cases explicitly not covered by object-level permissions, or perhaps some permissions not _yet_ supported by Teleport.

Reply:
TBH I think db_roles and db_permissions should be mutually exclusive. Otherwise it becomes very difficult to keep track of everything. Basically: you either manage your permissions in Postgres or in Teleport, not in both.

Reply:
I've added a note that the features should be mutually exclusive.